City-of-Helsinki/drupal-hdbt
# UHF-3643: Mobile navigation accessibility

# Mobile navigation accessibility [UHF-3643](https://helsinkisolutionoffice.atlassian.net/browse/UHF-3643)

The task was to fix the navigation problems in our current mobile menu. This proved out to be way harder than thought out. In essence, the whole mobile navigation code was rewritten.

## What was done

* Old mobile navigation library was removed
* New cssnav mobile navigation was created from scratch based on accessibility standards and recommendations
  * Works with javascript disabled or missing
  * Follows accessibility best practises (with some known limitations in pure-css form)
    * Checkbox and its label have `role="button"`for guiding screen readers to announce them as buttons
    * When menu is open, page content gets hidden with `display:none` to prevent focus leaking outside menu
    * Menu toggle button and sub-section buttons have screen reader accessible labels, menu text toggles between two modes
    * Menu toggle button is focusable (This was one of the original criterias)
    * Menu toggle button is focused when the menu is opened (even without js) (This was one of the original criterias)
    * Active link focus-like styles were added for better Safari browser support
    * When menu is open, all its contents are searchable with browser standard search and browser opens the menu item automatically (works on browsers that support this like chrome and edge)
  * Progressive enhancement when JS is enabled:
    * The link button switches from checkbox label to a real button that has aria-expanded attribute
      * Reduces logical focus problems with screen readers
      * Reduces screen reader confusion between navigation and form states
    * When the mobile menu is open, pressing ESC key closes the menu and focuses the menu button (One of the original criterias)
    * Javascript widgets like chats are forced hidden when the menu is open
* Header top navigation contents were introduced to mobile menu, so that also mobile users can navigate to hel.fi main homepage and adjanced links
* Menu opens and closes with a CSS animation from or towards the button to keep user focus flow logical
* Various language menu switcher fixes
  * Buggy language switcher animations on screen resize were fixed
  * Language menu switcher scrolling amount was set to automatic instead of hardcoded magic number
  * Code readability improvements
* Round borders from focus rings were removed on whole system as requested by design team
* Moved some of the drupal toolbar logic magic numbers to css variable (`--drupal-toolbar-height`) for easier referencing.
* Removed some logical problems on main menu breakpoint renderings (items were appearing twice in certaing widths)
* Fixed untranslated header top navigation title from current site
* Added translations for all new elements
* Limited menu levels to 5 when counting from "instance, main menu item, level 3, level 4, level 5" as requested by design team.
  * Code still supports menu levels up to 10 as does Drupal, so we can increase this number when there is need.

## How to install

* Make sure your instance is up and running on latest dev branch.
  * `git pull origin dev`
  * `make fresh`
* Update the HDBT theme
  * `composer require drupal/hdbt:dev-UHF-3643_mobile_nav_a11y`
* Run `make drush-cr`

## How to test

* Test mobile, between and desktop sizes
* Test keyboard focusing and navigating to mobile menu (check that other header components tabbing did not break either)
* Test mouse usage
* Test closing menu with esc key (with js enabled only)
* Test with and without javascript enabled
* Check menu levels at least to level 5 and above, anything above 5 should not be visible in mobile or sidebar menus, but still accessible in breadcrumb
* If you wish, you can test with some screen reader like VoiceOver, and/or you can check what chrome accessibility tools tell about the elements and their namings. The earlier code an concepts have been tested by accessibility team and this will be evaluated again by them. In no means should this be a downgrade from currently keyboard unusable production mobile menu.
